diff --git a/sysklogd/syslogd.c b/sysklogd/syslogd.c
index a119bde..275dd9b 100644
--- a/sysklogd/syslogd.c
+++ b/sysklogd/syslogd.c
@@ -824,8 +824,8 @@ static void timestamp_and_log(int pri, char *msg, int len)
 		time(&now);
 		timestamp = ctime(&now) + 4; /* skip day of week */
 	} else {
-		now = 0;
-		timestamp = msg;
+		time(&now);
+		timestamp = ctime(&now) + 4; /* skip day of week */
 		msg += 16;
 	}
 	timestamp[15] = '\0';
