Index: procd-2015-10-29.1/service/service.c
===================================================================
--- procd-2015-10-29.1.orig/service/service.c
+++ procd-2015-10-29.1/service/service.c
@@ -153,9 +153,9 @@ service_delete(struct service *s)
 	vlist_flush_all(&s->instances);
 	avl_delete(&services, &s->avl);
 	trigger_del(s);
+	service_validate_del(s);
 	free(s->trigger);
 	free(s);
-	service_validate_del(s);
 }
 
 enum {
