Index: procd-2015-10-29.1/service/instance.c
===================================================================
--- procd-2015-10-29.1.orig/service/instance.c
+++ procd-2015-10-29.1/service/instance.c
@@ -937,6 +937,7 @@ void instance_dump(struct blob_buf *b, s
 		blobmsg_add_u32(b, "threshold", in->respawn_threshold);
 		blobmsg_add_u32(b, "timeout", in->respawn_timeout);
 		blobmsg_add_u32(b, "retry", in->respawn_retry);
+		blobmsg_add_u32(b, "count", in->respawn_count);
 		blobmsg_close_table(b, r);
 	}
 
