Index: samba-3.6.25/librpc/idl/dcerpc.idl
===================================================================
--- samba-3.6.25.orig/librpc/idl/dcerpc.idl
+++ samba-3.6.25/librpc/idl/dcerpc.idl
@@ -468,7 +468,7 @@ interface dcerpc
 	/* Cancel was pending at sender */
 	const int DCERPC_PFC_FLAG_PENDING_CANCEL =
 		DCERPC_PFC_FLAG_PENDING_CANCEL_OR_HDR_SIGNING;
-	const ist DCERPC_PFC_FLAG_SUPPORT_HEADER_SIGN =
+	const int DCERPC_PFC_FLAG_SUPPORT_HEADER_SIGN =
 		DCERPC_PFC_FLAG_PENDING_CANCEL_OR_HDR_SIGNING;
 
 	/* these offsets are needed by the signing code */
