Index: wireless_tools.29/iwlist.c
===================================================================
--- wireless_tools.29.orig/iwlist.c
+++ wireless_tools.29/iwlist.c
@@ -85,6 +85,7 @@ typedef struct iw_auth_descr
 #define IW_SCAN_ONE_CHAN_ACTIVE   0x0800   /* Scan only one channel */
 #define IW_SCAN_FAST_DISPLAY   0x0200   /* Display fast scan result */
 #define IW_SCAN_FAST_DISPLAY_ADV   0x0400   /* Display fast scan advance result */
+#define IW_SCAN_FAST_DISPLAY_ALL   0x1000   /* Display fast scan normal+advance result */
 #endif
 
 #define IW_EXTKEY_SIZE	(sizeof(struct iw_encode_ext) + IW_ENCODING_TOKEN_MAX)
@@ -585,6 +586,23 @@ print_scanning_token(struct stream_descr
 #endif
       else if (state->type == AC)
         printf("11ac\t");
+
+#if 0//SENAO_SUPPORT_BACKGROUND_FASTSCAN
+      if(event->u.data.flags & SCAN_ADV_HT_FLAG)
+      {
+          char ht_option;
+          memcpy(&ht_option, event->u.data.pointer, event->u.data.length);
+
+          if(ht_option == SCAN_ADV_HT40_PLUS)
+              printf("HT40PLUS");
+          else if(ht_option == SCAN_ADV_HT40_MINUS)
+              printf("HT40MINUS");
+          else if(ht_option == SCAN_ADV_HT80)
+              printf("VHT80");
+          else
+              printf("HT20");
+      }
+#endif
       printf("\n%s\t", iw_saether_ntop(&event->u.ap_addr, buffer));
       state->ap_num++;
        state->sec = OPEN;
@@ -725,13 +743,13 @@ print_scanning_token(struct stream_descr
               memcpy(&ht_option, event->u.data.pointer, event->u.data.length);
 
               if(ht_option == SCAN_ADV_HT40_PLUS)
-                  printf("HT40PLUS");
+                  printf("%-10.10s\t", "HT40PLUS");
               else if(ht_option == SCAN_ADV_HT40_MINUS)
-                  printf("HT40MINUS");
+                  printf("%-10.10s\t","HT40MINUS");
               else if(ht_option == SCAN_ADV_HT80)
-                  printf("VHT80");
+                  printf("%-10.10s\t","VHT80");
               else
-                  printf("HT20");
+                  printf("%-10.10s\t","HT20");
           }
           else
           {
@@ -1164,6 +1182,7 @@ display_scanning_info(int     skfd,
   int           has_range;
   struct timeval    tv;             /* Select timeout */
   int advance_display = 0; /* SENAO advance display mode */
+  int all_display=0; /*SENAO all display mode*/
 #if ATH_SUPPORT_WRAP
   int           timeout = 60000000;     /* 60s */
 #else
@@ -1212,6 +1231,11 @@ display_scanning_info(int     skfd,
       count--;
 
     /* Display normal scanning result */
+    if(!strncmp(args[0], "all", 3))
+    {
+       all_display = 1;
+    }
+    else
     if(!strncmp(args[0], "normal", 4))
       {
         advance_display = 0;
@@ -1316,6 +1340,9 @@ display_scanning_info(int     skfd,
 
       /* SENAO fast scan display */
       wrq.u.data.flags |= IW_SCAN_FAST_DISPLAY;
+      if(all_display)
+	  wrq.u.data.flags |= IW_SCAN_FAST_DISPLAY_ALL;
+      else
       if(advance_display)
           wrq.u.data.flags |= IW_SCAN_FAST_DISPLAY_ADV;
 
@@ -1389,6 +1416,21 @@ display_scanning_info(int     skfd,
 #endif
 
       printf("%-8.16s  Scan completed :\n", ifname);
+      if(all_display)
+      {
+        printf("%-17.17s\t%-32.32s\t%-3.3s\t%-6.6s\t%-10.10s\t%-2.2s\t%-6.6s\t%-12.12s\t%-4.4s\t"
+              , "BSSID"
+              , "SSID"
+              , "LEN"
+              , "MODE"
+	      , "HT"
+              , "CHAN"
+              , "SIGNAL"
+              , "ENC"
+              , "TYPE"
+          );
+
+      }else
       if(advance_display)
       {
       printf("%-17.17s\t%-6.6s\t"
